src/machine/machined-varlink.c

```diff
@@ -507,6 +508,126 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
         return lookup_machine_and_call_method(link, parameters, flags, userdata, vl_method_terminate_internal);
 }
 
+static int list_image_one(sd_varlink *link, Image *image, bool more) {
+        int r;
+
+        assert(link);
+        assert(image);
+
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_STRING("name", image->name),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("path", image->path),
+                        SD_JSON_BUILD_PAIR("type", SD_JSON_BUILD_STRING(image_type_to_string(image->type))),
+                        SD_JSON_BUILD_PAIR("class", SD_JSON_BUILD_STRING(image_class_to_string(image->class))),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("readOnly", image->read_only),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("creationTimestamp", image->crtime),
+                        JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("modificationTimestamp", image->mtime),
+                        JSON_BUILD_PAIR_UNSIGNED_NOT_EQUAL("usage", image->usage, UINT64_MAX),
+                        JSON_BUILD_PAIR_UNSIGNED_NOT_EQUAL("usageExclusive", image->usage_exclusive, UINT64_MAX),
+                        JSON_BUILD_PAIR_UNSIGNED_NOT_EQUAL("limit", image->limit, UINT64_MAX),
+                        JSON_BUILD_PAIR_UNSIGNED_NOT_EQUAL("limitExclusive", image->limit_exclusive, UINT64_MAX),
+                        SD_JSON_BUILD_PAIR_CONDITION(image->metadata_valid && !!image->hostname, "hostname", SD_JSON_BUILD_STRING(image->hostname)),
+                        SD_JSON_BUILD_PAIR_CONDITION(image->metadata_valid && !sd_id128_is_null(image->machine_id),
+                                                     "machineId",
+                                                     SD_JSON_BUILD_ID128(image->machine_id)),
+                        SD_JSON_BUILD_PAIR_CONDITION(image->metadata_valid && image->machine_info,
+                                                     "machineInfo",
+                                                     JSON_BUILD_STRV_ENV_PAIR(image->machine_info)),
+                        SD_JSON_BUILD_PAIR_CONDITION(image->metadata_valid && image->os_release,
+                                                     "OSRelease",
+                                                     JSON_BUILD_STRV_ENV_PAIR(image->os_release)));
```

Comment:
doesn't matter, but instead of repeating the image->metadata_valid check here each time, you could also do that in a separate sd_json_variant_merge_objectbo() call.

i.e.:

```c
if (image->metadata_valid) {
        r = sd_json_variant_merge_objectbo(&v, SD_JSON_BUILD_PAIR_CONDITION(…));
        if (r < 0)
                return r;
}
```

but doesn't really matter. Might be a bit easier to read though